TypeSystem/flow-analysis/reachability_for_A01_t10.dart

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2025, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion for statement: If `N` is a for statement of the form
+/// `for (D; C; U) S,` then:
+/// - Let `before(D) = before(N)`.
+/// - Let `before(C) = conservativeJoin(after(D), assignedIn(N), capturedIn(N))`
+/// - Let `before(S) = split(true(C))`.
+/// - Let `before(U) = merge(after(S), continue(S))`.
+/// - Let `after(N) = inheritTested(join(false(C), unsplit(break(S))), after(U))`
+///
+/// @description Checks that
+/// `before(C) = conservativeJoin(after(D), assignedIn(N), capturedIn(N))`. Test
+/// that `capturedIn(C)` is detected by flow analysis.
+/// @author sgrekhov22@gmail.com
+
+int? x = (2 > 1) ? 1 : null;
+
+test1(int? n) {
+  if (n != null) { // n promoted to `int`
+    for (; () {
+        late int? i = (n = x); // n demoted to `int?`
```

Comment:
I don't think write-captured variables care about which type we're assigning to them: They can't be promoted at any location where the write-capture has occurred (that is: the function object exists, no matter whether we "know" that it has or hasn't been called). So `x` could be a plain `int`, and it would still prevent all promotion of `n`.

I think this implies that the test would be slightly stronger if `x` were actually of type `int`.

Reply:
Also, the flow analysis makes no attempt to prove that `n = x` is dead code because we never read the value of `i`. All we need is the fact that `n = x` _occurs_ in the function literal, no matter where.